app/src/main/java/woowacourse/movie/list/presenter/ReservationHistoryPresenter.kt

```diff
@@ -0,0 +1,21 @@
+package woowacourse.movie.list.presenter
+
+import androidx.fragment.app.Fragment
+import woowacourse.movie.database.TicketDatabase
+import woowacourse.movie.list.contract.ReservationHistoryContract
+import woowacourse.movie.ticket.model.TicketEntity
+import kotlin.concurrent.thread
+
+class ReservationHistoryPresenter(
+    private val view: ReservationHistoryContract.View,
+) : ReservationHistoryContract.Presenter {
+    private lateinit var tickets: List<TicketEntity>
+    override fun loadDataFromDb() {
+        val context = (view as Fragment).requireContext()
+        val ticketDb = TicketDatabase.getDatabase(context)
+        thread {
+            tickets = ticketDb.ticketDao().getAll()
+        }.join()
+        view.showData(tickets)
```

Comment:
이곳도 context에 대해 고민이 많으셨던 흔적이 있네요 ㅎㅎ..
이제 알송과 android의 context에 대해서 이야기를 나누면 이번 미션도 끝이 날 것 같네요.

대표적으로 안드로이드에는 application context 와 activity context가 존재합니다.
이 둘의 차이점은 다음 미션으로 넘어가기 전 꼭 찾아보고 학습 후 넘어가셨으면 해요.

이부분도 sharedPreference와 마찬가지로 모델로서, 외부로부터 주입받으면 좋을텐데요
어떻게해야 Presenter를 생성할 때 함께 Database를 주입받을 수 있을지 [글](https://developer.android.com/codelabs/android-room-with-a-view-kotlin?_gl=1*123trwt*_up*MQ..*_ga*NDc3MDYwMjQ1LjE3MTU1NzQ0MDc.*_ga_6HH9YJMN9M*MTcxNTU3NDQwNy4xLjAuMTcxNTU3NDQwNy4wLjAuMA..#12)을 읽어보고 고민해보세요!

정리
1. application context vs activity context 각가 정의와 차이점
2. sharedPreference, dataStore, room 에서는 어떤 context 의존하는게 좋을지

Reply:
1. Application Context는 앱의 전체 수명 동안 존재하는 context이며, 앱 전체의 전역적인 context를 나타냅니다. 애플리케이션이 시작될 때 생성되고, 종료될 때 사라집니다. 즉 하나의 애플리케이션에 하나만 존재합니다. 앱 전체의 전역적인 상태에 접근할 때 사용됩니다.
Activity context는 액티비티의 생명주기 동안에만 존재하는 context입니다. 여러 액티비티마다 각각 존재합니다. 액티비티가 create될 때 생성되고 destroy될 때 사라집니다. 해당 액티비티에서 수행되는 작업이나 액티비티의 상태에 접근할 때 사용됩니다.

2. shared Preference, dataStore, room의 경우 하나의 액티비티에서만 사용되는 것이 아닌, 앱 전체에서 사용됩니다. 따라서 앱 전체의 전역적인 상태에 접근해야 하므로 Application context를 사용해야 한다고 생각합니다.

지금까지 미션을 하면서 깊게 생각하지 못했던 부분인데, 제이든이 짚어주셔서 바르게 사용하고 넘어갈 수 있게 되었습니다! 정말 감사합니다 😁😆

주신 링크를 읽어보니, 프레젠터의 생성자로 db나 shared preference를 주입받기 위해 by lazy를 쓰는 것 같습니다. 그 이유는 액티비티가 create되기 전에는 activity context가 아직 만들어지지 않았기 때문에 application context에 접근하지 못하기 때문인 것 같습니다. 실제로 액티비티의 프로퍼티로 바로 `this.applicationContext`를 사용했더니 null을 반환하는 것을 알 수 있었습니다! fragment도 마찬가지로 onAttach()가 호출되기 전에는 에러가 발생했네요.

그리고 이것들을 공부하는 과정에서 들었던 의문점이 있는데요..!

1. by lazy가 아닌 lateinit var를 써도 무방할까요?

2. Application context도 결국 앱 프로세스가 종료되면 사라집니다. 그런데 shared preference, db같은 녀석들은 앱 프로세스가 종료되어도 남아있어야 합니다. 그렇다면 Application context로도 커버되지 않고, 더 넓은 범위를 관리하는 context가 필요한 것이 아닌가? 라는 생각이 얼핏 들었습니다. 물론 그런 것은 없겠지만요.. ㅋㅋ 그래서 어쩔 수 없이 application context를 쓸 수밖에 없는 것일까요?

이번 미션동안 많이 배웠습니다! 제이든과 함께하면 정말 크게 성장하는 느낌이 드네요. 다시한번 감사드립니다!! 😘

Reply:
> 프레젠터의 생성자로 db나 shared preference를 주입받기 위해 by lazy를 쓰는 것 같습니다. 

이유에 대해서는 글에도 나와있는데요, 앱을 실행할 때 생성하기 보다 처음 호출될 때 생성하기 위해 lazy가 사용되었습니다.
`Because these objects should only be created when they're first needed, rather than at app startup, you're using Kotlin's property delegation: by lazy.`

1. 사실 lateinit을 사용해서 [onCreate](https://developer.android.com/reference/android/app/Application#onCreate()) 시점에 초기화를 해주어도 무방하겠지만, 위 설명처럼 호출될 때 초기화를 위해 lazy가 보다 목적에 알맞지 않을까 생각이 드네요!

2. 왜 ApplicationContext에 접근을 하려면 activity를 통해야 하는지는 Application 과 Activity는 어떤 관계이며, 둘의 context는 어떻게 생성이 되고 소멸되는지 학습하시다보면 자연스럽게 해소될 부분입니다 ㅎㅎ 
또는, [Application](https://developer.android.com/reference/android/app/Application)의 상속관계도 살펴보시면 도움이 될겁니다.

3. sharedPreferences와 room 은 디바이스의 물리적인 저장소 영역에 데이터를 저장하기 위한 수단입니다. 이 수단을 이용하기 위해 앱 프로세스가 종료되어도 남아있는 context가 필요할까요..?
물리적인 데이터는 context와 관련없이 존재하는 요소입니다. 이곳을 접근하기 위한 context 통로가 필요하여 사용하는게 applicationContext가 되는 것이죠.